src/main/scala/vectorpipe/osm/internal/ElementToFeature.scala

```diff
@@ -271,46 +275,42 @@ private[vectorpipe] object ElementToFeature {
    * Note: The [[State]] monad usage here is for counting Lines which couldn't
    * be fused.
    */
-  private def fuseLines(v: Vector[OSMLine]): State[Int, Vector[OSMPolygon]] = v match {
-    case Vector() => Vector.empty.pure[IntState]
-    case v if v.length == 1 => State.modify[Int](_ + 1).map(_ => Vector.empty)
-    case v => {
-      fuseOne(v).flatMap({
-        case None => fuseLines(v.tail)
-        case Some((f, d, i, rest)) => {
-          if (f.isClosed)
-            fuseLines(rest).map(c => Feature(Polygon(f), d) +: c)
-          else {
-            /* The next sections of the currently accumulating Line may be far down
-             * the Vector, so we punt it forward to avoid wasted traversals
-             * and intersection checks.
-             * This is a band-aid around the fact that `spatialSort` doesn't
-             * place all sections of every fusable line in order; clumps of them
-             * can appear far apart in the Vector.
-             *
-             * If `i == 0`, the splitAt and `(++)` below are basically no-ops,
-             * so there's no point in any `if (i == 0) skipTheSplitAt`.
-             */
-            val (a, b) = rest.splitAt(i)
-
-            fuseLines(a ++ (Feature(f, d) +: b))
-          }
+  private def fuseLines(v: Vector[OSMLine]): State[List[OSMLine], Vector[OSMPolygon]] = v match {
+    case Vector()  => Vector.empty.pure[LineState]
+    case Vector(h) => State.modify[List[OSMLine]](h :: _).map(_ => Vector.empty)
+    case v => fuseOne(v) match {
+      case None => State.modify[List[OSMLine]](v.head :: _) >> fuseLines(v.tail)
+      case Some((f, d, i, rest)) => {
+        if (f.isClosed)
+          fuseLines(rest).map(c => Feature(Polygon(f), d) +: c)
+        else {
+          /* The next sections of the currently accumulating Line may be far down
+           * the Vector, so we punt it forward to avoid wasted traversals
+           * and intersection checks.
+           * This is a band-aid around the fact that `spatialSort` doesn't
+           * place all sections of every fusable line in order; clumps of them
+           * can appear far apart in the Vector.
+           *
+           * If `i == 0`, the splitAt and `(++)` below are basically no-ops,
+           * so there's no point in any `if (i == 0) skipTheSplitAt`.
+           */
+          val (a, b) = rest.splitAt(i)
+
+          fuseLines(a ++ (Feature(f, d) +: b))
         }
-      })
+      }
     }
   }
 
   /**
    * Fuse the head Line in the Vector with the first other Line possible.
    * This borrows [[fuseLines]]'s assumptions.
    */
-  private def fuseOne(
-    v: Vector[OSMLine]
-  ): State[Int, Option[(Line, ElementData, Int, Vector[OSMLine])]] = {
+  private def fuseOne(v: Vector[OSMLine]): Option[(Line, ElementData, Int, Vector[OSMLine])] = {
```

Comment:
Was able to make this pure again. The `State` stuff is handled back up in `fuseLines`.